src/lavinmqperf/mqtt/throughput.cr

```diff
@@ -0,0 +1,345 @@
+require "../../lavinmq/mqtt/protocol"
+require "json"
+require "wait_group"
+require "../perf"
+
+module LavinMQPerf
+  module MQTT
+    class Throughput < Perf
+      @publishers = 1
+      @consumers = 1
+      @size = 16
+      @verify = false
+      @topic = "perf-test"
+      @qos = 0
+      @rate = 0
+      @consume_rate = 0
+      @quiet = false
+      @json_output = false
+      @timeout = Time::Span.zero
+      @pmessages = 0
+      @cmessages = 0
+      @random_bodies = false
+      @retain = false
+      @clean_session = false
+      @uri = URI.parse("mqtt://localhost:1883")
+
+      def initialize
+        super
+        @parser.on("-x publishers", "--publishers=number", "Number of publishers (default 1)") do |v|
+          @publishers = v.to_i
+        end
+        @parser.on("-y consumers", "--consumers=number", "Number of consumers (default 1)") do |v|
+          @consumers = v.to_i
+        end
+        @parser.on("-s msgsize", "--size=bytes", "Size of each message (default 16 bytes)") do |v|
+          @size = v.to_i
+        end
+        @parser.on("-V", "--verify", "Verify the message body") do
+          @verify = true
+        end
+        @parser.on("-q qos", "--qos=level", "QoS level (0 or 1)") do |v|
+          @qos = v.to_i
+        end
+        @parser.on("-t topic", "--topic=name", "Topic name (default perf-test)") do |v|
+          @topic = v
+        end
+        @parser.on("-r pub-rate", "--rate=number", "Max publish rate (default 0)") do |v|
+          @rate = v.to_i
+        end
+        @parser.on("-R consumer-rate", "--consumer-rate=number", "Max consume rate (default 0)") do |v|
+          @consume_rate = v.to_i
+        end
+        @parser.on("-j", "--json", "Output result as JSON") do
+          @json_output = true
+        end
+        @parser.on("-z seconds", "--time=seconds", "Only run for X seconds") do |v|
+          @timeout = Time::Span.new(seconds: v.to_i)
+        end
+        @parser.on("-q", "--quiet", "Quiet, only print the summary") do
+          @quiet = true
+        end
+        @parser.on("-C messages", "--pmessages=messages", "Publish max X number of messages") do |v|
+          @pmessages = v.to_i
+        end
+        @parser.on("-D messages", "--cmessages=messages", "Consume max X number of messages") do |v|
+          @cmessages = v.to_i
+        end
+        @parser.on("--random-bodies", "Each message body is random") do
+          @random_bodies = true
+        end
+        @parser.on("--retain", "Set retain flag on published messages") do
+          @retain = true
+        end
+        @parser.on("--clean-session", "Use clean session") do
+          @clean_session = true
+        end
+        @parser.on("-u uri", "--uri=uri", "MQTT broker URI (default mqtt://localhost:1883)") do |v|
+          @uri = URI.parse(v)
```

Comment:
Not sure `URI.parse` is the best here, it fails quietly when given somethign it doesn't understand. 

For example: 
`-u mqtt://hostname:8822` works but since the protocol isn't used is tempting to write just `-u hostname:8822` but that doesn't work and it defaults to localhost:1883. 

If we want the scheme to matter, one option is to remove the protocol check in LavinMQPerf and only look at what schema is used for the URI. defaulting to `amqp://localhhost` as we did before. This also gives us backwards compatibility.

Reply:
If i understand correctly, then you would always need to specify a uri if you want to run lavinmqperf with mqtt? 
so just running locally that would be
for amqp: lavinmqperf throughput
but for mqtt: lavinmqperf throguhtput --uri=mqtt://localhost:1883

Reply:
I took inspo from the omq rabbitmq client 
https://github.com/cloudamqp/lavinmq/issues/971#issuecomment-2695720805

Reply:
Yes, that would be the result. 
Benefits are that this change will be backward compatible but since you already discussed this I'm fine with the current change. Makes it nice to run it locally.

Reply:
But back to the issue, URI.parse since it fails silently. 
I like that when I enter a badly formatted argument, the program tells me whats wrong instead of using defaults and just continue. Is there a `URI.validate` or something we can use? 
Or just raise on line 88 instead of `|| "localhost"`  to tell that something is wrong.

Reply:
yeah I get what you mean.. I'm not sure either what is best. First i thought its not really something we integrate anywhere, more of a local tool, but now that we use it with the benchmarking terraform tool its a different story

Reply:
```
 u = URI.parse "localhost:1883"
 => #<URI:0x7f799d52f9c0 @scheme="localhost", @host=nil, @port=nil, @path="1883", @query=nil, @user=nil, @password=nil, @fragment=nil>
```

you can have some logic when `u.host == u.port == nil` maybe?

Reply:
```crystal 
if u.host == u.port == nil
  u.host = u.scheme
  u.scheme = "mqtt"
  u.port = u.path.to_i
  u.path = nil
end

Reply:
I added a conditional reformatting if there is no uri host/port, so now it should help a user when they enter `host:port` 
I alos added a log line printing out what uri the connection has used so its clear, idk @mange if you think thats enough or do you still think it should raise instead of defaulting to localhost?

Reply:
I added a conditional reformatting if there is no uri host/port, so now it should help a user when they enter `host:port` 
I alos added a log line printing out what uri the connection has used so its clear, idk @snichme  if you think thats enough or do you still think it should raise instead of defaulting to localhost?